### ISSUE TITLE
Inline comments on Github

### ISSUE THREAD

**Issue by champo** [danger/danger#346]
While Danger can leave a comment on an PR with lint messages in a table, it would be much more useful to comment on the offending line. I understand the [GH API](https://developer.github.com/v3/pulls/comments/#create-a-comment) is limited and only allows comments on lines included in the commit diff. Any solution using this API should include some sort of fallback to using a table when the commented lines are not included in the diff.

Is there any reason danger doesn't support this? 

## Progress
 - [x] Add access to the diff (https://github.com/danger/danger/pull/355)
 - [x] Modify the messaging plugin to support file and line (https://github.com/danger/danger/pull/356)
 - [x] Show file and line information in the comment table (#371) (#373)
 - [x] Add inline comments using GH API (#412)

**Comment by orta** [comment]
No one has looked at doing it, I'd love to see that happen personally 👍

**Comment by KrauseFx** [comment]
I'm also curios on how we can make this easy to use from the `Dangerfile` without adding too much complexity.

**Comment by orta** [comment]
I'd imagine we have two ways to add to it:

The first one is _failable_, it would work by downloading the diffs, looking through all of them to see if that file is referenced, then doing a lookup for the line position to the hunk position.

```ruby
github.add_comment_to_file("CHANGELOG.md", line: 30, comment: "Please add a space after the line")
```

The other one involves reading through the diff itself and making assumptions based off that

```ruby
diff = github.pr_diff
if diff.string.include? "fdescribe"
  line = diff.line_for_first "fdescribe"
  github.add_comment_to_diff(diff, line: line, comment: "Don't include an fdescribe in your tests")
end
```

**Comment by champo** [comment]
@orta I was thinking something along the same lines. The first alternative would call the second one when possible, since GH can only handle comments on the diff itself and not on the file.
My idea was to transparently fallback to adding a table of "inline" comments like some of the example Dangerfiles do now.

**Comment by orta** [comment]
Yeah - that sounds perfect to me

**Comment by champo** [comment]
@orta I think this might end being a massive PR. I'm probably going to split it in several smaller steps if I can.

**Comment by orta** [comment]
I think that's a great idea, I'd imagine the simplest starting point would be to start building something to grab the diff hunks when requested, then exposing them as  `diff = github.pr_diff` in some form

**Comment by JuanitoFatas** [comment]
Implemented in #412.
